common/core/desktop/tests/unit/kmnkbd/test_kmx_context.cpp

```diff
@@ -0,0 +1,414 @@
+  assert_string_equal(buf, u"bcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkl");
+}
+
+void
+test_Add() {
+  KMX_Context context;
+  KMX_WCHAR buf[2 * MAXCONTEXT];
+
+  context.Add(u'a');
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"a");
+
+  context.Reset();
+  context.Add(0xD801);
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_equal(u16len(buf), 1);
+  assert_equal(buf[0], 0xD801);
+
+  context.Reset();
+  context.Add(0xD801);
+  context.Add(0xDC12);
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"\U00010412");
+
+  context.Reset();
+  context.Add(u'\uFFFF');
+  context.Add(u'\u0008');
+  context.Add(u'\u0001');
+  context.Add(u'a');
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"\uFFFF\u0008\u0001a");
+
+  context.Reset();
+  context.Set(u"\uFFFF\u0008\u0001abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefgh");
+  context.Add(u'1');
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefgh1");
+}
+
+void
+test_Delete() {
+  KMX_Context context;
+  KMX_WCHAR buf[2 * MAXCONTEXT];
+
+  context.Reset();
+  context.Delete();
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"");
+
+  context.Reset();
+  context.Set(u"abc");
+  context.Delete();
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"ab");
+
+  context.Reset();
+  context.Set(u"ab\U00010412");
+  context.Delete();
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"ab");
+
+  context.Reset();
+  context.Set(u"ab\uFFFF\u0008\u0001");
+  context.Delete();
+  context.Get((KMX_WCHAR *)&buf, 2 * MAXCONTEXT);
+  assert_string_equal(buf, u"ab");
+}
+
+void
+test_Buf() {
+  KMX_Context context;
+
+  context.Reset();
+  context.Set(u"abc");
+  auto last = context.Buf(0);
+  auto p    = context.Buf(1);
+  assert_equal(p, last - 1);
+
+  p = context.Buf(2);
+  assert_equal(p, last - 2);
+
+  p = context.Buf(3);
+  assert_equal(p, last - 3);
+
+  p = context.Buf(4);
+  assert(p == NULL);
+
+  p = context.Buf(10);
+  assert(p == NULL);
+
+  p = context.Buf(-1);
+  assert_equal(p, last);
+
+  context.Reset();
+  context.Set(u"ab\U00010412");
+  last = context.Buf(0);
+  p    = context.Buf(1);
+  assert_equal(p, last - 2);
+
+  p = context.Buf(2);
+  assert_equal(p, last - 3);
+
+  p = context.Buf(3);
+  assert_equal(p, last - 4);
+
+  p = context.Buf(4);
+  assert(p == NULL);
+
+  p = context.Buf(5);
+  assert(p == NULL);
+
+  p = context.Buf(10);
+  assert(p == NULL);
+
+  context.Reset();
+  context.Set(u"ab\uFFFF\u0008\u0001");
+  last = context.Buf(0);
+  p    = context.Buf(1);
+  assert_equal(p, last - 3);
+
+  p = context.Buf(2);
+  assert_equal(p, last - 4);
+
+  p = context.Buf(3);
+  assert_equal(p, last - 5);
+
+  p = context.Buf(4);
+  assert(p == NULL);
```

Comment:
I can see why this in the `Buf` code why this returns NULL. However, is this a bug - is this the intent of `Buf` when the caller asks for more valid xstring units than is present in the `CurContext`?

Reply:
`Buf` is a weird one ... for historical reasons, that's what it does -- it was for performance during rule evaluation, saying, well if I can't even get enough characters to compare, I don't need to do the comparison. Function name is obviously awful!

Reply:
Can I suggest we update the doxy explanation in `kmx_context.h` to include this explanation.  I can do it as a separate LHF PR. as all the `@brief`  tags need to be removed from that header fine as well.

Reply:
> Can I suggest we update the doxy explanation in kmx_context.h to include this explanation. I can do it as a separate LHF PR. as all the @brief tags need to be removed from that header fine as well.

Sounds good to me!